drivers/spi/spi_ll_stm32.c

```diff
@@ -1,36 +1,171 @@
+#define SPI_STM32_SPI_1_IRQ_PRI \
+	_SPI_STM32_SPI_1_IRQ_PRI1(SOC_SPI_CONTROLLER_0)
+#define _SPI_STM32_SPI_1_IRQ_PRI1(_prefix) \
+	_SPI_STM32_SPI_1_IRQ_PRI2(_prefix)
+#define _SPI_STM32_SPI_1_IRQ_PRI2(_prefix) \
+	_SPI_STM32_SPI_1_IRQ_PRI3(_prefix##_IRQ_0_PRIORITY)
+#define _SPI_STM32_SPI_1_IRQ_PRI3(_irq_pri) _irq_pri
+/* SPI 2 */
+#define SPI_STM32_SPI_2			SPI_STM32_SPI_2_BASE_ADDRESS
+#define SPI_STM32_SPI_2_NAME \
+	_SPI_STM32_SPI_2_NAME1(SOC_SPI_CONTROLLER_1)
+#define _SPI_STM32_SPI_2_NAME1(_prefix) \
+	_SPI_STM32_SPI_2_NAME2(_prefix)
+#define _SPI_STM32_SPI_2_NAME2(_prefix) \
+	_SPI_STM32_SPI_2_NAME3(_prefix##_LABEL)
+#define _SPI_STM32_SPI_2_NAME3(_name) _name
+#define SPI_STM32_SPI_2_BASE_ADDRESS \
+	_SPI_STM32_SPI_2_BASE_ADDRESS1(SOC_SPI_CONTROLLER_1)
+#define _SPI_STM32_SPI_2_BASE_ADDRESS1(_prefix) \
+	_SPI_STM32_SPI_2_BASE_ADDRESS2(_prefix)
+#define _SPI_STM32_SPI_2_BASE_ADDRESS2(_prefix) \
+	_SPI_STM32_SPI_2_BASE_ADDRESS3(_prefix##_BASE_ADDRESS)
+#define _SPI_STM32_SPI_2_BASE_ADDRESS3(_base_address) _base_address
+#define SPI_STM32_SPI_2_IRQ \
+	_SPI_STM32_SPI_2_IRQ1(SOC_SPI_CONTROLLER_1)
+#define _SPI_STM32_SPI_2_IRQ1(_prefix) \
+	_SPI_STM32_SPI_2_IRQ2(_prefix)
+#define _SPI_STM32_SPI_2_IRQ2(_prefix) \
+	_SPI_STM32_SPI_2_IRQ3(_prefix##_IRQ_0)
+#define _SPI_STM32_SPI_2_IRQ3(_irq) _irq
+#define SPI_STM32_SPI_2_IRQ_PRI \
+	_SPI_STM32_SPI_2_IRQ_PRI1(SOC_SPI_CONTROLLER_1)
+#define _SPI_STM32_SPI_2_IRQ_PRI1(_prefix) \
+	_SPI_STM32_SPI_2_IRQ_PRI2(_prefix)
+#define _SPI_STM32_SPI_2_IRQ_PRI2(_prefix) \
+	_SPI_STM32_SPI_2_IRQ_PRI3(_prefix##_IRQ_0_PRIORITY)
+#define _SPI_STM32_SPI_2_IRQ_PRI3(_irq_pri) _irq_pri
+/* SPI 1 */
+#define SPI_STM32_SPI_3			SPI_STM32_SPI_3_BASE_ADDRESS
+#define SPI_STM32_SPI_3_NAME \
+	_SPI_STM32_SPI_3_NAME1(SOC_SPI_CONTROLLER_2)
+#define _SPI_STM32_SPI_3_NAME1(_prefix) \
+	_SPI_STM32_SPI_3_NAME2(_prefix)
+#define _SPI_STM32_SPI_3_NAME2(_prefix) \
+	_SPI_STM32_SPI_3_NAME3(_prefix##_LABEL)
+#define _SPI_STM32_SPI_3_NAME3(_name) _name
+#define SPI_STM32_SPI_3_BASE_ADDRESS \
+	_SPI_STM32_SPI_3_BASE_ADDRESS1(SOC_SPI_CONTROLLER_2)
+#define _SPI_STM32_SPI_3_BASE_ADDRESS1(_prefix) \
+	_SPI_STM32_SPI_3_BASE_ADDRESS2(_prefix)
+#define _SPI_STM32_SPI_3_BASE_ADDRESS2(_prefix) \
+	_SPI_STM32_SPI_3_BASE_ADDRESS3(_prefix##_BASE_ADDRESS)
+#define _SPI_STM32_SPI_3_BASE_ADDRESS3(_base_address) _base_address
+#define SPI_STM32_SPI_3_IRQ \
+	_SPI_STM32_SPI_3_IRQ1(SOC_SPI_CONTROLLER_2)
+#define _SPI_STM32_SPI_3_IRQ1(_prefix) \
+	_SPI_STM32_SPI_3_IRQ2(_prefix)
+#define _SPI_STM32_SPI_3_IRQ2(_prefix) \
+	_SPI_STM32_SPI_3_IRQ3(_prefix##_IRQ_0)
+#define _SPI_STM32_SPI_3_IRQ3(_irq) _irq
+#define SPI_STM32_SPI_3_IRQ_PRI \
+	_SPI_STM32_SPI_3_IRQ_PRI1(SOC_SPI_CONTROLLER_2)
+#define _SPI_STM32_SPI_3_IRQ_PRI1(_prefix) \
+	_SPI_STM32_SPI_3_IRQ_PRI2(_prefix)
+#define _SPI_STM32_SPI_3_IRQ_PRI2(_prefix) \
+	_SPI_STM32_SPI_3_IRQ_PRI3(_prefix##_IRQ_0_PRIORITY)
+#define _SPI_STM32_SPI_3_IRQ_PRI3(_irq_pri) _irq_pri
```

Comment:
@b0661: One additional question. Would it be possible with chaos library to generate instances code on the fly, w/o need for defining this whole bunch of macros per instance?
This seems somewhat like a burden to provide all these, I wonder how it will be easy for most people (me for instance) to do the same work for their driver. At least if it could be minimized to a generic part and not per instance, it would be easier.
For instance, STM32PWM could have around 20 instances, this will be a big nasty file...

Reply:
@erwango 
> Would it be possible with chaos library to generate instances code on the fly, w/o need for defining this whole bunch of macros per instance?

Yes it is. I did not change that to keep the driver the way it is as much as possible. The new version of the driver (not yet pushed) just needs the following:

    /* Driver Settings from device tree
     */
    #define SPI_STM32_SPI_COUNT	SOC_SPI_CONTROLLER_COUNT
    #define SPI_STM32_SPI_0		SOC_SPI_CONTROLLER_0
    #define SPI_STM32_SPI_1		SOC_SPI_CONTROLLER_1
    #define SPI_STM32_SPI_2		SOC_SPI_CONTROLLER_2
 
The SOC_SPI_CONTROLLER_x macros are generated for all activated SPI nodes. The value of the macro is the respective prefix (e.g. ST_STM32_SPI_FIFO_40000000). 

> This seems somewhat like a burden to provide all these, I wonder how it will be easy for most people (me for instance) to do the same work for their driver.

You are right, there should be some convenience/ helper macros to ease the driver implementers work and prevent the faults I made. I'm thinking of something like:

    #define DRIVER_PROPERTY(prefix, idx, property)

This may be used to retrieve device tree properties easily:

    spi_stm32_0_base_address = DRIVER_PROPERTY(SPI_STM32_SPI, 0, BASE_ADDRESS)

I will add a driver_lib.h to gather such driver implementation support macros and functions.

> For instance, STM32PWM could have around 20 instances, this will be a big nasty file...

For the PWM driver also SOC_PWM_CONTROLLER_x macros can be generated. Currently that requires a pwm-controller directive in the device tree. This is not inline with the Linux "standard" binding. There are [device]-controller directives for gpio, clock, pin(ctrl), interrupt device types. I think it would make sense to introduce this for all device types. Otherwise the device type has to be deduced from e.g. the compatible directive. This would need some heuristics.

@erwango, in  #3487 you set the goal to keep Linux and Zephyr device tree specifications in sync. Do you think a [device]-controller directive can be added to Zephyr device tree specifications fo all relevant devices?

Reply:
> Do you think a [device]-controller directive can be added to Zephyr device tree specifications fo all relevant devices?

@b0661 : Yes, I think it is relevent. Maybe @galak @agross-linaro could confirm